File: android/liveness/src/main/java/com/liveness/detection/TransformationMatrixExtractor.kt
package com.liveness.detection

import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarkerResult

object TransformationMatrixExtractor {
  fun extract(result: FaceLandmarkerResult): FloatArray? {
    return try {
      val method = result.javaClass.getMethod("facialTransformationMatrixes")
      val matrices = method.invoke(result) as? List<*>
      val first = matrices?.firstOrNull() ?: return null
      extractMatrixData(first)
    } catch (_: Exception) {
      null
    }
  }

  private fun extractMatrixData(matrixObject: Any): FloatArray? {
    val candidates = listOf("getDataList", "dataList", "getData", "data")
    for (name in candidates) {
      try {
        val method = matrixObject.javaClass.getMethod(name)
        val data = method.invoke(matrixObject)
        val list = data as? List<*>
        if (list != null && list.isNotEmpty()) {
          val floats = list.mapNotNull { it as? Float }
          if (floats.isNotEmpty()) return floats.toFloatArray()
        }
      } catch (_: Exception) {
        continue
      }
    }
    return null
  }
}

File: ios/LivenessDetection/TransformationMatrixExtractor.swift
import Foundation
import MediaPipeTasksVision

enum TransformationMatrixExtractor {
  static func extract(from result: FaceLandmarkerResult) -> [Float]? {
    let mirror = Mirror(reflecting: result)
    for child in mirror.children {
      if child.label == "facialTransformationMatrixes" {
        return flattenMatrix(child.value)
      }
    }
    return nil
  }

  private static func flattenMatrix(_ value: Any) -> [Float]? {
    if let array = value as? [[Float]], let first = array.first {
      return first
    }
    if let array = value as? [Float] {
      return array
    }
    if let array = value as? [Any], let first = array.first {
      if let data = first as? [Float] {
        return data
      }
      if let mirror = Mirror(reflecting: first).children.first(where: { $0.label == "data" }) {
        return mirror.value as? [Float]
      }
    }
    return nil
  }
}
